clang/lib/AST/ByteCode/Compiler.cpp

```diff
@@ -4368,8 +4363,19 @@ template <class Emitter> bool Compiler<Emitter>::visitIfStmt(const IfStmt *IS) {
     if (!visitDeclStmt(CondDecl))
       return false;
 
-  if (!this->visitBool(IS->getCond()))
-    return false;
+  // Compile condition.
+  if (IS->isNonNegatedConsteval()) {
+    if (!this->emitIsConstantContext(IS))
+      return false;
+  } else if (IS->isNegatedConsteval()) {
+    if (!this->emitIsConstantContext(IS))
+      return false;
+    if (!this->emitInv(IS))
+      return false;
+  } else {
+    if (!this->visitBool(IS->getCond()))
+      return false;
+  }
```

Comment:
We really should not evaluate the condition, just the underlying statement directly
(getThen or getElse depending on both `S.inConstantContext()` and whether it's negated)

Reply:
The condition is only evaluated if the ifstmt is not a consteval stmt at all

Reply:
I mean, we should not have a `emitIsConstantContext(IS)` - just emit the underlying statement

Reply:
That's what the previous code did, though. That doesn't work if we compile to bytecode and later interpret the same bytecode in a non-constant context.

Reply:
How would you deal with code like

```cpp
constexpr void f() {
    if !consteval {
goto a;
a:;  
}
}```

Reply:
Just emitting ops for invalid code won't produce diagnostics, so that's not a problem. It would only be problematic if the thenstmt in this case caused the compilation to be aborted.

Reply:
I understand  your concern but I think this is conceptually no different than
```c++
constexpr int f(bool b) {
if (!b) {
  goto a;
  a:;
} else {
}
  return  1;
}
```
where the current interpreter would also only evaluate one of the branches while the bytecode interpreter needs to generate code for both.

Reply:
@cor3ntin Ping

Reply:
I guess my question is whether doing that would prevent us to evaluate/create bytecode for statements for which the non-taken branch has things the byte code interpreter can't compile yet.
Including non-constexpr function calls, UB, goto, coroutines, exceptions, asm blocks, builtins, simd, threads, etc etc

Reply:
Only if that has truly not been implemented yet. Otherwise, we need to always diagnose it anyway, in the same way we do for regular if statements.

Reply:
the whole point of `if consteval` is to allow a runtime evaluation to do thing that would not be valid in constant context, such as asm or simd or reinterpret cast or crazy vendor extensions.

so 
```cpp
constexpr void f() {
if !consteval {
  // copy a volatile vla with simd and execute a gpu kernel
}
}
```

is intended to always work.
I don't think it would be reasonable to delay adoption of the new interpreter until we support _all_ the features of clang in the bytecode.
So this approach is probably not viable.
If you really want to create bytecode for both branches, you need to have an escape hatch so that if the non-consteval branch cannot produce bytecode, we still can support the consteval branch without error

Reply:
Stuff like that _needs_ to be "implemented" in the interpreter. It must be, because it must be rejected as well. It's the same example as the `goto` statement above. It needs to be compiled to bytecode, somehow. If it isn't then using it in a constexpr function (without the `if consteval` will also fail).

Again, this is no different than any other if statement. I could imagine compiling two different versions of functions (one for consteval = true, one for =false), but we have the same problem with other if statements. This is just a difference between the interpreting and compiling approach.

Reply:
Ok, it took me a while, but iI think i got it... sorry for the brain fart.
Because the interpreter compile functions before they are evaluated (at least morally), we don't know in which context they are going to be executed.

This is somewhat confusing because that means which assume we can constant evaluate something that is not a constant expression, which i guess is true for constant folding. 

In that light the patch makes sense, I think

I still wants to make sure that we can recover nicely in the presence of things we can't evaluate, which seems to be the case https://godbolt.org/z/bofa9b8de


Do you do the same thing for `is_constant_evaluated`?

Reply:
> Do you do the same thing for `is_constant_evaluated`?

Yes:
https://github.com/llvm/llvm-project/blob/933f72217e4584db03f945a3b30e8c04537f4dab/clang/lib/AST/ByteCode/InterpBuiltin.cpp#L162